### ISSUE TITLE
Bug in validation

### ISSUE THREAD

**Issue by carlpe** [Project-MONAI/MONAI#3098]
Greetings,

I have found a bug in during validation of images. I believe it is something with validation transforms in the latest versions of Monai.  Everything works fine on Docker version 0.5.0 (projectmonai/monai:0.5.0) , but fails on Docker version => 0.7.0 and latest.

I ran the exact same validation code on both versions and here is the output of validation dataset shape.

It seems shapes are different.

v0.5.0: torch.Size([1, 14, 336, 336, 336]) torch.Size([1, 14, 336, 336, 336]) 
v0.7.0: torch.Size([14, 14, 336, 336, 336]) torch.Size([14, 1, 336, 336, 336])

**Comment by carlpe** [comment]
import os
import glob
import shutil
import tempfile
import monai
from monai.transforms.croppad.dictionary import RandSpatialCropD, RandSpatialCropd
import torch
from monai.utils import first, set_determinism
import matplotlib.pyplot as plt
import numpy as np
from monai.networks.nets import UNet
from monai.data import CacheDataset, DataLoader, Dataset, NiftiSaver, list_data_collate
from tqdm import tqdm
from monai.losses import DiceCELoss,DiceLoss
from monai.inferers import sliding_window_inference
from monai.transforms import (AsDiscrete,AddChanneld,Activations,Rand3DElasticd,Compose,SpatialCropd,ToTensor,CropForegroundd,LoadImaged,Resized,RandScaleIntensityd,
    Orientationd,RandFlipd,RandCropByPosNegLabeld,RandShiftIntensityd,NormalizeIntensityd,ScaleIntensityRanged,Spacingd,RandRotate90d,ToTensord,)
from monai.config import print_config
from monai.metrics import DiceMetric
from monai.metrics import compute_meandice, get_confusion_matrix
#from monai.networks.nets import UNETR
from monai.data import (DataLoader,CacheDataset,SmartCacheDataset, decollate_batch,)
from torch.utils.tensorboard import SummaryWriter 

device=("cuda:0")
torch.cuda.is_available()
#monai.config.print_debug_info()
#torch.cuda.get_device_name(0)

# DIRECTORIES
#directory = os.environ.get(r"/MONAI_DATA_DIRECTORY/")
#root_dir = r"D:\Documents\monai\tutorials\runet" if directory is None else directory
#path_to_target = r"D:\Documents\niftynet\data\mect"

directory = os.environ.get('/MONAI_DATA_DIRECTORY/')
root_dir = '/titan/disko/monai/' if directory is None else directory
#path_to_target = '/titan/disko/mect/data/'
path_to_target = "/titan/disko/data/dlk/"

# IMAGES
train_images = sorted(glob.glob(os.path.join(path_to_target, "train", "*T1.nii")))
tgt_images = sorted(glob.glob(os.path.join(path_to_target, "traingt", "*.nii.gz")))
val_images = sorted(glob.glob(os.path.join(path_to_target, "val", "*T1.nii")))
valgt_images = sorted(glob.glob(os.path.join(path_to_target, "valgt", "*.nii.gz")))
#print(len(train_images), len(tgt_images), len(val_images), len(valgt_images))   

# DICTIONARIES
train_dicts = [{"image": image_name, "label": label_name}  for image_name, label_name in zip(train_images[:20], tgt_images[:20])]
val_dicts = [{"image": image_name, "label": label_name} for image_name, label_name in zip(train_images[:5], tgt_images[:5])]
print("Training ][ Validation")
print(len(train_dicts), len(val_dicts))

# TRANSFORMS
train_transforms = Compose([
        LoadImaged(keys=["image", "label"]),
        AddChanneld(keys=["image", "label"]),
        #Resized(keys=['image', 'label'], spatial_size=(336,336,336)),
        RandScaleIntensityd(keys=["image"], factors = 0.1, prob=0.3),  
        #RandSpatialCropd(keys=["image", "label"],roi_size=[256,256,256],random_size=False),
        #Spacingd(keys=["image", "label"],pixdim=(1, 1, 1),mode=("bilinear", "nearest"),),
        RandCropByPosNegLabeld(keys=["image", "label"], label_key="label", spatial_size=[336,336,336], pos=1, neg=1, num_samples=1),
        #Rand3DElasticd(keys=["image", "label"], mode=("bilinear", "nearest"), prob=0.5, sigma_range=(5, 8), magnitude_range=(50, 100), 
        #    spatial_size=(256,256,256), scale_range=(0.02, 0.02, 0.02), padding_mode="border"),
        RandFlipd(keys=["image", "label"], prob=0.4, spatial_axis=[0,1,2]),
        NormalizeIntensityd(keys=['image'], nonzero=True, channel_wise=True),
        ToTensord(keys=["image", "label"]),])

val_transforms = Compose([
        LoadImaged(keys=["image", "label"]),
        AddChanneld(keys=["image", "label"]),
        #RandSpatialCropd(keys=["image", "label"],roi_size=[336,336,336],random_size=False),
        #Spacingd(keys=["image", "label"],pixdim=(1, 1, 1),mode=("bilinear", "nearest"),),
        #Resized(keys=['image', 'label'], spatial_size=(336,336,336)),
        RandCropByPosNegLabeld(keys=["image", "label"], label_key="label", spatial_size=[336,336,336], pos=1, neg=1, num_samples=1),  
        NormalizeIntensityd(keys=['image'], nonzero=True, channel_wise=True),
        ToTensord(keys=["image", "label"]),])

# DETERMINISM
set_determinism(seed=0)

# LOAD DATA
train_ds = CacheDataset(data=train_dicts, transform=train_transforms,cache_num=20,cache_rate=1.0,num_workers=4)
#train_ds = SmartCacheDataset(data=train_dicts, transform=train_transforms,cache_rate=1.0,replace_rate=0.2,cache_num=39,num_init_workers=2,num_replace_workers=2)
train_loader = DataLoader(train_ds, batch_size=1, shuffle=True)
val_ds = Dataset(data=val_dicts, transform=val_transforms)
val_loader = DataLoader(val_ds, batch_size=1)
#check_data = first(train_loader)
#print(check_data["image"].shape, check_data["label"].shape)

# ?
post_trans = Compose([Activations(sigmoid=True), AsDiscrete(threshold_values=True)])

# DEVICE
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

# NETWORK
model = UNet(dimensions=3,in_channels=1,out_channels=14,channels=(16,32,64,128,256),strides=(2,2,2,2),num_res_units=2).to(device)
loss_function = DiceLoss(to_onehot_y=True,softmax=True)
optimizer = torch.optim.Adam(model.parameters(), 0.0001)

# LOAD MODEL
#model.load_state_dict(torch.load("/titan/disko/monai/unet.pth"))
print("Model loaded")

val_interval = 1
best_metric = -1
best_metric_epoch = -1
epoch_loss_values = list()
metric_values = list()
post_pred = AsDiscrete(argmax=True, to_onehot=True, n_classes = 14)
post_label = AsDiscrete(to_onehot=True, n_classes = 14)

writer = SummaryWriter()

# TRAINING
for epoch in range(200):
    print("-" * 10)
    print(f"epoch {epoch + 1}/{200}")
    model.train()
    epoch_loss = 0
    step = 0
    for batch_data in train_loader:
        step += 1
        inputs, labels = batch_data["image"].to(device), batch_data["label"].to(device)
        optimizer.zero_grad()
        outputs = model(inputs)
        loss = loss_function(outputs, labels)
        loss.backward()
        optimizer.step()
        epoch_loss += loss.item()
        epoch_len = len(train_ds) // train_loader.batch_size
        print(f"{step}/{epoch_len}, Train_loss: {loss.item():.4f}")
        writer.add_scalar("Train_loss", loss.item(), epoch_len * epoch + step)
    epoch_loss /= step
    epoch_loss_values.append(epoch_loss)
    print(f"Epoch {epoch + 1} Average loss: {epoch_loss:.4f}")

    
# VALIDATION
    if (epoch + 1) % val_interval == 0:
        model.eval()
        with torch.no_grad(): 
            metric_sum = 0.0
            metric_count = 0
            #val_images = None
            #val_labels = None
            #val_outputs = None
            for val_data in val_loader:
                val_images, val_labels = val_data["image"].to(device), val_data["label"].to(device)
                roi_size = (336,336,336)
                sw_batch_size = 1
                #val_outputs = sliding_window_inference(val_images, roi_size, sw_batch_size, model)
                val_outputs = model(val_images)
                #print(val_outputs.shape)
                val_outputs = post_pred(val_outputs)
                val_labels = post_label(val_labels)
                print(val_outputs.shape,val_labels.shape)
                value = compute_meandice( y_pred=val_outputs, y=val_labels, include_background=False,)
                metric_count += len(value)
                metric_sum += value.sum().item()    #value.item() * len(value)

            metric = metric_sum / metric_count
            metric_values.append(metric)
            if metric > best_metric:
                best_metric = metric
                best_metric_epoch = epoch + 1
                torch.save(model.state_dict(), "dlk.pth")
                print("saved new best metric model")
            print("current epoch: {} current mean dice: {:.4f} best mean dice: {:.4f} at epoch {}".format(
                    epoch + 1, metric, best_metric, best_metric_epoch))
            writer.add_scalar("val_mean_dice", metric, epoch + 1)  
            print(f"train completed, best_metric: {best_metric:.4f} at epoch: {best_metric_epoch}") 

"""

"""

**Comment by carlpe** [comment]
```
import os
import glob
import shutil
import tempfile
import monai
from monai.transforms.croppad.dictionary import RandSpatialCropD, RandSpatialCropd
import torch
from monai.utils import first, set_determinism
import matplotlib.pyplot as plt
import numpy as np
from monai.networks.nets import UNet
from monai.data import CacheDataset, DataLoader, Dataset, NiftiSaver, list_data_collate
from tqdm import tqdm
from monai.losses import DiceCELoss,DiceLoss
from monai.inferers import sliding_window_inference
from monai.transforms import (AsDiscrete,AddChanneld,Activations,Rand3DElasticd,Compose,SpatialCropd,ToTensor,CropForegroundd,LoadImaged,Resized,RandScaleIntensityd,
    Orientationd,RandFlipd,RandCropByPosNegLabeld,RandShiftIntensityd,NormalizeIntensityd,ScaleIntensityRanged,Spacingd,RandRotate90d,ToTensord,)
from monai.config import print_config
from monai.metrics import DiceMetric
from monai.metrics import compute_meandice, get_confusion_matrix
#from monai.networks.nets import UNETR
from monai.data import (DataLoader,CacheDataset,SmartCacheDataset, decollate_batch,)
from torch.utils.tensorboard import SummaryWriter 

device=("cuda:0")
torch.cuda.is_available()
#monai.config.print_debug_info()
#torch.cuda.get_device_name(0)

# DIRECTORIES
#directory = os.environ.get(r"/MONAI_DATA_DIRECTORY/")
#root_dir = r"D:\Documents\monai\tutorials\runet" if directory is None else directory
#path_to_target = r"D:\Documents\niftynet\data\mect"

directory = os.environ.get('/MONAI_DATA_DIRECTORY/')
root_dir = '/titan/disko/monai/' if directory is None else directory
#path_to_target = '/titan/disko/mect/data/'
path_to_target = "/titan/disko/data/dlk/"

# IMAGES
train_images = sorted(glob.glob(os.path.join(path_to_target, "train", "*T1.nii")))
tgt_images = sorted(glob.glob(os.path.join(path_to_target, "traingt", "*.nii.gz")))
val_images = sorted(glob.glob(os.path.join(path_to_target, "val", "*T1.nii")))
valgt_images = sorted(glob.glob(os.path.join(path_to_target, "valgt", "*.nii.gz")))
#print(len(train_images), len(tgt_images), len(val_images), len(valgt_images))   

# DICTIONARIES
train_dicts = [{"image": image_name, "label": label_name}  for image_name, label_name in zip(train_images[:20], tgt_images[:20])]
val_dicts = [{"image": image_name, "label": label_name} for image_name, label_name in zip(train_images[:5], tgt_images[:5])]
print("Training ][ Validation")
print(len(train_dicts), len(val_dicts))

# TRANSFORMS
train_transforms = Compose([
        LoadImaged(keys=["image", "label"]),
        AddChanneld(keys=["image", "label"]),
        #Resized(keys=['image', 'label'], spatial_size=(336,336,336)),
        RandScaleIntensityd(keys=["image"], factors = 0.1, prob=0.3),  
        #RandSpatialCropd(keys=["image", "label"],roi_size=[256,256,256],random_size=False),
        #Spacingd(keys=["image", "label"],pixdim=(1, 1, 1),mode=("bilinear", "nearest"),),
        RandCropByPosNegLabeld(keys=["image", "label"], label_key="label", spatial_size=[336,336,336], pos=1, neg=1, num_samples=1),
        #Rand3DElasticd(keys=["image", "label"], mode=("bilinear", "nearest"), prob=0.5, sigma_range=(5, 8), magnitude_range=(50, 100), 
        #    spatial_size=(256,256,256), scale_range=(0.02, 0.02, 0.02), padding_mode="border"),
        RandFlipd(keys=["image", "label"], prob=0.4, spatial_axis=[0,1,2]),
        NormalizeIntensityd(keys=['image'], nonzero=True, channel_wise=True),
        ToTensord(keys=["image", "label"]),])

val_transforms = Compose([
        LoadImaged(keys=["image", "label"]),
        AddChanneld(keys=["image", "label"]),
        #RandSpatialCropd(keys=["image", "label"],roi_size=[336,336,336],random_size=False),
        #Spacingd(keys=["image", "label"],pixdim=(1, 1, 1),mode=("bilinear", "nearest"),),
        #Resized(keys=['image', 'label'], spatial_size=(336,336,336)),
        RandCropByPosNegLabeld(keys=["image", "label"], label_key="label", spatial_size=[336,336,336], pos=1, neg=1, num_samples=1),  
        NormalizeIntensityd(keys=['image'], nonzero=True, channel_wise=True),
        ToTensord(keys=["image", "label"]),])

# DETERMINISM
set_determinism(seed=0)

# LOAD DATA
train_ds = CacheDataset(data=train_dicts, transform=train_transforms,cache_num=20,cache_rate=1.0,num_workers=4)
#train_ds = SmartCacheDataset(data=train_dicts, transform=train_transforms,cache_rate=1.0,replace_rate=0.2,cache_num=39,num_init_workers=2,num_replace_workers=2)
train_loader = DataLoader(train_ds, batch_size=1, shuffle=True)
val_ds = Dataset(data=val_dicts, transform=val_transforms)
val_loader = DataLoader(val_ds, batch_size=1)
#check_data = first(train_loader)
#print(check_data["image"].shape, check_data["label"].shape)

# ?
post_trans = Compose([Activations(sigmoid=True), AsDiscrete(threshold_values=True)])

# DEVICE
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

# NETWORK
model = UNet(dimensions=3,in_channels=1,out_channels=14,channels=(16,32,64,128,256),strides=(2,2,2,2),num_res_units=2).to(device)
loss_function = DiceLoss(to_onehot_y=True,softmax=True)
optimizer = torch.optim.Adam(model.parameters(), 0.0001)

# LOAD MODEL
#model.load_state_dict(torch.load("/titan/disko/monai/unet.pth"))
print("Model loaded")

val_interval = 1
best_metric = -1
best_metric_epoch = -1
epoch_loss_values = list()
metric_values = list()
post_pred = AsDiscrete(argmax=True, to_onehot=True, n_classes = 14)
post_label = AsDiscrete(to_onehot=True, n_classes = 14)

writer = SummaryWriter()

# TRAINING
for epoch in range(200):
    print("-" * 10)
    print(f"epoch {epoch + 1}/{200}")
    model.train()
    epoch_loss = 0
    step = 0
    for batch_data in train_loader:
        step += 1
        inputs, labels = batch_data["image"].to(device), batch_data["label"].to(device)
        optimizer.zero_grad()
        outputs = model(inputs)
        loss = loss_function(outputs, labels)
        loss.backward()
        optimizer.step()
        epoch_loss += loss.item()
        epoch_len = len(train_ds) // train_loader.batch_size
        print(f"{step}/{epoch_len}, Train_loss: {loss.item():.4f}")
        writer.add_scalar("Train_loss", loss.item(), epoch_len * epoch + step)
    epoch_loss /= step
    epoch_loss_values.append(epoch_loss)
    print(f"Epoch {epoch + 1} Average loss: {epoch_loss:.4f}")

    
# VALIDATION
    if (epoch + 1) % val_interval == 0:
        model.eval()
        with torch.no_grad(): 
            metric_sum = 0.0
            metric_count = 0
            #val_images = None
            #val_labels = None
            #val_outputs = None
            for val_data in val_loader:
                val_images, val_labels = val_data["image"].to(device), val_data["label"].to(device)
                roi_size = (336,336,336)
                sw_batch_size = 1
                #val_outputs = sliding_window_inference(val_images, roi_size, sw_batch_size, model)
                val_outputs = model(val_images)
                #print(val_outputs.shape)
                val_outputs = post_pred(val_outputs)
                val_labels = post_label(val_labels)
                print(val_outputs.shape,val_labels.shape)
                value = compute_meandice( y_pred=val_outputs, y=val_labels, include_background=False,)
                metric_count += len(value)
                metric_sum += value.sum().item()    #value.item() * len(value)

            metric = metric_sum / metric_count
            metric_values.append(metric)
            if metric > best_metric:
                best_metric = metric
                best_metric_epoch = epoch + 1
                torch.save(model.state_dict(), "dlk.pth")
                print("saved new best metric model")
            print("current epoch: {} current mean dice: {:.4f} best mean dice: {:.4f} at epoch {}".format(
                    epoch + 1, metric, best_metric, best_metric_epoch))
            writer.add_scalar("val_mean_dice", metric, epoch + 1)  
            print(f"train completed, best_metric: {best_metric:.4f} at epoch: {best_metric_epoch}") 

"""

"""

```

**Comment by carlpe** [comment]
Thank you @Nic-Ma ! It works now. 
![image](https://user-images.githubusercontent.com/54416001/136773641-6b8a54cd-d351-4255-820d-d97ff2b1065b.png)


I do have another question for you if you do not mind me asking here.  In this example I  load in T1-images exclusively. However, I have three more MR-weightings (channels) also in the same folder.. named:

/train/DLK001_T1
/train/DLK001_PD
/train/DLK001_FS
/train/DLK001_ANG

/traingt/DLK001_GT
/traingt/DLK002_GT



Could you please explain to me how I load all four channels to the right labelfile?  so 4 input niftivolumes and 1 GT  for each subject. 

I suppose this is a rather simple problem
